### DISCUSSION TITLE
Dependency cycle

### DISCUSSION THREAD

**Discussion by tcharding** [bitcoindevkit/bdk#1226] [Q&A]
I wanted to see the dependency graph for BDK so I could see what order things need upgrading when we do a `rust-bitcoin` release. Here is what I came up with, it looks like there is a dependency cycle when considering the upgrade path (ie., the cycle only exists when considering dev-dependencies. How do you guys normally get around this?

`electrsd` -> `bitcoin_slices` -> `bitcoin-test-data` -> `electrsd`

EDIT: The line above is incorrect, it was accidentally using the (now removed) dependency graph of `electrs` not `electrsd`

# Dependency graph for the bdk stack of crates

Dependencies controlled by either the BDK team or the rust-bitcoin team.

Includes crates by RCasatta:

- electrsd
- bitcoin-test-data
- bitcoin_slices

## electrsd

electrsd v0.26.0
├── bitcoind v0.33.3
└── electrum-client v0.15.1

[build-dependencies]
└── bitcoin_hashes v0.12.0 (*)

## bitcoin_slices

bitcoin_slices v0.7.0
├── bitcoin v0.31.0
└── bitcoin_hashes v0.13.0 (*)

[dev-dependencies]
├── bitcoin v0.31.0 (*)
├── bitcoin-test-data v0.2.0
└── hex_lit v0.1.1

## bitcoin-test-data 

bitcoin-test-data v0.2.0
[dev-dependencies]
└── electrsd v0.23.3

## rust-esplora-client

esplora-client v0.6.0
├── bitcoin v0.30.2
└── bitcoin-internals v0.1.0

[dev-dependencies]
├── electrsd v0.24.0
└── electrum-client v0.16.0

## rust-electrum-client

electrum-client v0.18.0
├── bitcoin v0.30.1
└── bitcoin-private v0.1.0

## rust-hwi

hwi v0.7.0
├── bitcoin v0.30.2
└── miniscript v10.0.0

### Crates

- `bdk`

bdk v1.0.0-alpha.2
├── bdk_chain v0.6.0
├── bip39 v1.2.0
├── bitcoin v0.30.1 (*)
├── hwi v0.7.0
└── miniscript v10.0.0 (*)

[dev-dependencies]
└── bdk_file_store v0.2.0

- `chain`

bdk_chain v0.6.0
├── bitcoin v0.30.1
└── miniscript v10.0.0

- `file_store`

bdk_file_store v0.2.0
└── bdk_chain v0.6.0

- `electrum`

bdk_electrum v0.4.0
├── bdk_chain
└── electrum-client v0.18.0

- `esplora`

bdk_esplora v0.4.0
├── bdk_chain v0.6.0
├── bitcoin v0.30.1 (*)
└── esplora-client v0.6.0

[dev-dependencies]
└── electrsd v0.25.0

- `bitcoind_rpc`

bdk_bitcoind_rpc v0.1.0
├── bdk_chain
├── bitcoin v0.30.1 (*)
└── bitcoincore-rpc v0.17.0

[dev-dependencies]
└── bitcoind v0.33.2

### Example crates

- `example_cli`

- `example electrum`

- `example_esplora`

- `example_bitcoind_rpc_polling`

- `wallet_electrum`

- `wallet_esplora_blocking`

- `wallet_esplora_async`

### Nursery

- `tmp_plan`

- `coin_select`

**Comment by tcharding** [comment]
In case you want to see the one for the `rust-bitcoin` crates, its here: https://github.com/rust-bitcoin/rust-bitcoin/discussions/2216

**Comment by apoelstra** [comment]
dev-dependencies are only compiled when testing a particular crate. You never compile the dev-dependencies of your dependencies.

**Reply by tcharding** [reply]
I was wrong, there is no cycle but not for the reason you stated. I had the graph for `electrs` not `electrsd` - that is not the first time I've mixed those two crates up.

**Comment by tcharding** [comment]
I was wrong, closing. I've updated the original post to use the dependency graph of `electrsd`
